decision-records/2025-03-17--otelcol-arbitrary-config-adr.md

```diff
@@ -0,0 +1,341 @@
+
+The OpenTelemetry config file has 5 main sections:
+
+
+| Section     | Description                                                                   |
+| ----------- | ----------------------------------------------------------------------------- |
+| `receivers` | To define where datasources are configured (logs, metrics, traces).           |
+| `processors`| To modify, filter or transform the data before sending them.                  |
+| `exporters` | To define where telemetry is going to be sent. (Loki, Prometheus, OTLP, etc.).|
+| `service`   | To define the pipelines which connects receivers, processors and exporters.   |
+| `extensions`| To configure extra features like health checks, authentication, encoding, etc.|
+
+
+The `receivers` section could be auto-generated based on the established relations, for example if we implement `cos-agent` interface we could generate something like this:
+
+```yaml
+receivers:
+  prometheus:
+    config:
+      scrape_configs:
+        - job_name: "my-app"
+          static_configs:
+            - targets: ["localhost:9090"]
+
+  # To read logs from disk
+  filelog:
+    include: ["/var/log/my-app.log"]
+    start_at: beginning
+    include_file_path: true
+    include_file_name: false
+    operators:
+      - type: regex_parser
+        regex: '^(?P<timestamp>\S+) (?P<log_level>\S+) (?P<message>.*)$'
+        timestamp:
+          parse_from: timestamp
+          layout: "%Y-%m-%dT%H:%M:%S.%fZ"
+
+  # To receive logs
+  loki:
+    protocols:
+      http:
+        endpoint: 0.0.0.0:3500
+    use_incoming_timestamp: true
+
+  otlp:
+    protocols:
+      grpc:
+        endpoint: 0.0.0.0:4317
+      http:
+        endpoint: 0.0.0.0:4318
+```
+
+
+The `exporters` sections could be ato-generated also based on the relations established, for example:
+
+```yaml
+exporters:
+  prometheus:
+    endpoint: "http://prometheus-server:9090/metrics"
+
+  loki:
+    endpoint: "http://loki-server:3100/loki/api/v1/push"
+
+  tempo:
+    endpoint: "http://tempo-server:3200/api/traces"
+```
+
+
+Finally the `service` section will relate `receivers` and `exporters`
+
+```yaml
+service:
+  pipelines:
+    metrics:
+      receivers: [prometheus]
+      exporters: [prometheus]
+
+    logs:
+      receivers: [loki, filelog]
+      exporters: [loki]
+
+    traces:
+      receivers: [otlp]
+      exporters: [tempo]
+```
+
+
+With configs like those ones we can replicate grafana-agent behaviour. Now we need to decide how to add `processors` and `extensions` to the config fiel.
+
+
+## Alternative 1: `juju config`
+
+In order to include `processors` or `extensions` we may use `juju config`, like this:
+
+```shell
+juju config otel-col processors_file='@path/to/processors-config.yaml' to_pipelines='metrics'
```

Comment:
If the processor-config has more than one processor, then a config option like `to_pipelines='metrics'` would probably not be sufficient to assign them to different pipelines?

Not pretty, but how about:
```bash
juju config pipelines_mapping=metricsgenerator=metrics,logssummarizer=logs`
```

Reply:
@MichaelThamm and I were talking earlier today about this in our pair-programming session... we have introduced a modification by removing the `to_pipelines` option in favour of a list in the file.

Let me know your thoughts!